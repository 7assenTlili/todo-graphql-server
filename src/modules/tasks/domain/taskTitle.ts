import {Guard} from "../../../shared/guard";
import config from "../../../config";

interface ITaskTitleProps {
  value: string
}

export class TaskTitle {
  public props: ITaskTitleProps
  public static minLength: number = config.MIN_TASK_TITLE_LENGTH;
  public static maxLength: number = config.MAX_TASK_TITLE_LENGTH;

  constructor(props: ITaskTitleProps) {
    this.props = props;
  }

  get value (): string {
    return this.props.value
  }

  public static create(name: string): TaskTitle {
    const { success: isValidMinLength } = Guard.againstMinLength(name, this.minLength);
    const { success: isValidMaxLength } = Guard.againstMaxLength(name, this.maxLength);
    if (!isValidMinLength) {
      throw new Error(`INVALID_USERNAME: should be at least ${this.minLength} characters`);
    } else if (!isValidMaxLength) {
      throw new Error(`INVALID_USERNAME: should be at least ${this.maxLength} characters`);
    } else {
      return new TaskTitle({value: name});
    }
  }
}
