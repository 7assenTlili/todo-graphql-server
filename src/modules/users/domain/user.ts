import { prop } from "@typegoose/typegoose";
import { ObjectId } from "mongodb";

export class User {
  readonly _id?: ObjectId;

  @prop()
  username: string;

  @prop()
  email: string;

  @prop()
  password: string;
}
