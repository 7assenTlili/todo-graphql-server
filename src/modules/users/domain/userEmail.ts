import {Guard} from "../../../shared/guard";

interface IUserEmailProps {
  value: string;
}

export class UserEmail {
  public props: IUserEmailProps

  constructor(props: IUserEmailProps) {
    this.props = props;
  }

  get value () : string {
    return this.props.value;
  }

  private static format(email: string): string {
    return email.trim().toLowerCase();
  }

  public static create(email: string): UserEmail {
    const { success: isValidEmail } = Guard.againstInvalidEmail(email)
    if (!isValidEmail) {
      throw new Error("INVALID_EMAIL")
    } else {
      return new UserEmail({value: this.format(email)});
    }
  }
}
