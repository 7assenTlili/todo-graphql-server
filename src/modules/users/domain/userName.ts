// @ts-ignore
import {Guard} from "../../../shared/guard";
import config from "../../../config";

interface IUserNameProps {
  value: string
}

export class UserName {
  public props: IUserNameProps
  public static minLength: number = config.MIN_USERNAME_LENGTH;
  public static maxLength: number = config.MAX_USERNAME_LENGTH;

  constructor(props: IUserNameProps) {
    this.props = props;
  }

  get value (): string {
    return this.props.value
  }

  public static create(name: string): UserName {
    const { success: isValidMinLength } = Guard.againstMinLength(name, this.minLength);
    const { success: isValidMaxLength } = Guard.againstMaxLength(name, this.maxLength);
    if (!isValidMinLength) {
      throw new Error(`INVALID_USERNAME: should be at least ${this.minLength} characters`);
    } else if (!isValidMaxLength) {
      throw new Error(`INVALID_USERNAME: should be at least ${this.maxLength} characters`);
    } else {
      return new UserName({value: name});
    }
  }
}
