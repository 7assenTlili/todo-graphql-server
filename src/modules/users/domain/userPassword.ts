// @ts-ignore
import * as bcrypt from "bcrypt-nodejs";
import {Guard} from "../../../shared/guard";
import config from "../../../config";

interface IUserPasswordProps {
  value: string
}

export class UserPassword {
  public props: IUserPasswordProps
  public static minLength: number = config.MIN_PASSWORD_LENGTH;

  constructor(props: IUserPasswordProps) {
    this.props = props;
  }

  public getHash(): Promise<string> {
    return new Promise((resolve, reject) => {
      bcrypt.hash(this.props.value, null, null, (err: any, hash: string) => {
        if (err) return reject(err);
        return resolve(hash);
      })
    })
  }

  private bcryptCompare (plainText: string, hashed: string): Promise<boolean> {
    return new Promise((resolve, reject) => {
      bcrypt.compare(plainText, hashed, (err: any, compareResult: boolean) => {
        if (err) return resolve(false);
        return resolve(compareResult);
      })
    })
  }

  public async compare(hash: string): Promise<boolean> {
    return this.bcryptCompare(this.props.value, hash);
  }

  public static create(password: string): UserPassword {
    const { success: isValidLength } = Guard.againstMinLength(password, this.minLength);
    if (!isValidLength) {
      throw new Error(`INVALID_PASSWORD: should be at least ${this.minLength} characters`);
    } else {
      return new UserPassword({value: password});
    }
  }
}
