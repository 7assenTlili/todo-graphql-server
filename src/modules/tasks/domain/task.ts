import {prop, Ref} from "@typegoose/typegoose";
import {User} from "../../users/domain/user";

export class Task {
  readonly _id?: string;

  @prop()
  title: string;

  @prop()
  description: string;

  @prop({ default: false })
  isComplete?: boolean;

  @prop({ ref: "User" })
  user?: Ref<User>

  @prop({ ref: "User" })
  sharedWith?: Ref<User>[]
}
