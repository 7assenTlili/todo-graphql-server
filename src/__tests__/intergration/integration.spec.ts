import { createTestClient } from "apollo-server-testing";
import gql from "graphql-tag";
import constructTestServer from "./constructTestServer";
import { closeDatabase, clearDatabase, connect, populateDatabase } from "../e2e/dbUtils";
import {getModelForClass} from "@typegoose/typegoose";
import {Task} from "../../modules/tasks/domain/task";
import {User} from "../../modules/users/domain/user";

beforeAll(async () => {
  await connect()

  const TaskModel = getModelForClass(Task);
  const UserModel = getModelForClass(User);

  await populateDatabase(TaskModel, [
    {
      _id: "60b0c3bc8cb1ea51949339e0",
      title: "abcd",
      description: "description test",
      user: "60b09157c813ff0f9c3b5569",
      sharedWith: [],
      isComplete: false
    },
    {
      _id: "60b0e0bcc672e602ac2f705e",
      title: "task2",
      description: "description test",
      user: "60a0c7d67883212b4c0f6bcc",
      sharedWith: [],
      isComplete: false
    },
    {
      _id: "60b0f43e23e5984df8a6fb8e",
      title: "task 4",
      description: "description test",
      user: "60a0c7d67883212b4c0f6bcc",
      sharedWith: ["60b09157c813ff0f9c3b5569"],
      isComplete: false
    },
    {
      _id: "60b15f1ee030c4362cdd7ff8",
      title: "to delete",
      description: "description test",
      user: "60b09157c813ff0f9c3b5569",
      sharedWith: [],
      isComplete: false
    },
  ]);

  await populateDatabase(UserModel, [
    {
      _id: "60b09157c813ff0f9c3b5569",
      email: "email@email.com",
      username: "hassen",
      password: "ncsjdcjs"
    },
    {
      _id: "60a0c7d67883212b4c0f6bcc",
      email: "user2@email.com",
      username: "user2",
      password: "ncsjdcjs"
    },
  ]);
});

afterAll(async (done) => {
  await clearDatabase();
  await closeDatabase();
  done();
});

describe("Tasks", () => {
  it("should return a user's own task", async () => {
    const server = constructTestServer({ user: {
      _id: "60b09157c813ff0f9c3b5569"
    }});

    // @ts-ignore
    const { query } = createTestClient(server);

    const GET_TASK = gql`
        {
            task(_id:"60b0c3bc8cb1ea51949339e0") {
                _id
                title
                user {
                    _id
                    email
                    username
                }
                sharedWith
            }
        }
    `;
    const res = await query({
      query: GET_TASK
    });

    expect(res).toMatchSnapshot();
  });

  it('should not return other user\'s task', async () => {
    const server = constructTestServer({ user: {
        _id: "60b09157c813ff0f9c3b5569"
      }});

    // @ts-ignore
    const { query } = createTestClient(server);

    const GET_TASK = gql`
        {
            task(_id:"60b0e0bcc672e602ac2f705e") {
                _id
                title
                user {
                    _id
                    email
                    username
                }
                sharedWith
            }
        }
    `;
    const res = await query({
      query: GET_TASK
    });

    expect(res).toMatchSnapshot();
  });

  it('should return a shared task', async () => {
    const server = constructTestServer({ user: {
        _id: "60b09157c813ff0f9c3b5569"
      }});

    // @ts-ignore
    const { query } = createTestClient(server);

    const GET_TASK = gql`
        {
            task(_id:"60b0f43e23e5984df8a6fb8e") {
                _id
                title
                user {
                    _id
                    email
                    username
                }
                sharedWith
            }
        }
    `;
    const res = await query({
      query: GET_TASK
    });

    expect(res).toMatchSnapshot();
  });

  it('should be able to create a task', async () => {
    const server = constructTestServer({ user: {
        _id: "60b09157c813ff0f9c3b5569"
      }});

    // @ts-ignore
    const { query } = createTestClient(server);

    const GET_TASK = gql`
        mutation {
            createTask(input: {
                title: "new Task"
                description:"description"
            }) {
                _id
                title
                description
                isComplete
                user {
                    _id
                }
                sharedWith
            }
        }
    `;
    const res = await query({
      query: GET_TASK
    });

    expect(res).toMatchSnapshot();
  });

  it('should be able to delete his own task', async () => {
    const server = constructTestServer({ user: {
        _id: "60b09157c813ff0f9c3b5569"
      }});

    // @ts-ignore
    const { query } = createTestClient(server);

    const GET_TASK = gql`
        mutation {
            deleteTask(id: "60b15f1ee030c4362cdd7ff8") {
                _id
                user {
                    email
                }
            }
        }
    `;
    const res = await query({
      query: GET_TASK
    });

    expect(res).toMatchSnapshot();
  });

  it('should not be able to delete shared task', async () => {
    const server = constructTestServer({ user: {
        _id: "60b09157c813ff0f9c3b5569"
      }});

    // @ts-ignore
    const { query } = createTestClient(server);

    const GET_TASK = gql`
        mutation {
            deleteTask(id: "60b0f43e23e5984df8a6fb8e") {
                _id
                user {
                    email
                }
            }
        }
    `;
    const res = await query({
      query: GET_TASK
    });

    expect(res).toMatchSnapshot();
  });
});
