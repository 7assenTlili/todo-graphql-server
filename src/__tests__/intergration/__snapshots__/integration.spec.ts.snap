// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks should be able to create a task 1`] = `
Object {
  "data": Object {
    "createTask": Object {
      "_id": "60b1ec4f057d7a313c738e92",
      "description": "description",
      "isComplete": false,
      "sharedWith": Array [],
      "title": "new Task",
      "user": Object {
        "_id": "60b09157c813ff0f9c3b5569",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tasks should be able to delete his own task 1`] = `
Object {
  "data": Object {
    "deleteTask": Object {
      "_id": "60b15f1ee030c4362cdd7ff8",
      "user": Object {
        "email": "email@email.com",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tasks should not be able to delete shared task 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: NOT AUTHORIZED],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tasks should not return other user's task 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: NOT_AUTHORIZED],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tasks should return a shared task 1`] = `
Object {
  "data": Object {
    "task": Object {
      "_id": "60b0f43e23e5984df8a6fb8e",
      "sharedWith": Array [
        "60b09157c813ff0f9c3b5569",
      ],
      "title": "task 4",
      "user": Object {
        "_id": "60a0c7d67883212b4c0f6bcc",
        "email": "user2@email.com",
        "username": "user2",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tasks should return a user's own task 1`] = `
Object {
  "data": Object {
    "task": Object {
      "_id": "60b0c3bc8cb1ea51949339e0",
      "sharedWith": Array [],
      "title": "abcd",
      "user": Object {
        "_id": "60b09157c813ff0f9c3b5569",
        "email": "email@email.com",
        "username": "hassen",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
